llama.cpp

```diff
@@ -14648,6 +14787,36 @@ LLAMA_API int32_t llama_chat_apply_template(
     return res;
 }
 
+LLAMA_API int llama_split_path(char * split_path, size_t maxlen, const char * path_prefix, int split_no, int split_count) {
+    static const char * const SPLIT_PATH_FORMAT = "%s-%05d-of-%05d.gguf";
+    if (snprintf(split_path, maxlen, SPLIT_PATH_FORMAT, path_prefix, split_no + 1, split_count)) {
+        return strlen(split_path);
+    }
+    return 0;
+}
+
+LLAMA_API int llama_split_prefix(char * dest, const char * split_path, size_t split_path_len, int split_no, int split_count) {
```

Comment:
The `split_path_len` maybe redundant here, because it can be calculated using `strlen(split_path)`

I also propose another version of this function, based on this pseudo code:

```
expected_postfix = format("-%05d-of-%05d.gguf", split_no, split_count)
if (split_path.endsWith(expected_postfix)):
  return split_path.substring(0, strlen(split_path) - strlen(expected_postfix))
else:
  return NULL
```

The cpp code:

```cpp
LLAMA_API int llama_split_prefix(char * dest, const char * split_path, int split_no, int split_count) {
    char * expected_postfix[128];
    sprintf(expected_postfix, "-%05d-of-%05d.gguf", split_no, split_count);

    const size_t path_len = strlen(path_len);
    const size_t postfix_len = strlen(expected_postfix);
    const char * offset = strstr(split_path, expected_postfix);

    // check if dest ends with expected_postfix
    if (offset && path_len > postfix_len && offset == split_path + path_len - postfix_len) {
        strncpy(dest, split_prefix, path_len - postfix_len);
        return path_len - postfix_len;
    }

    return 0;
}
```

Reply:
This is a common practice to pass the string len in parameter. Anyway, it will not pass the compiler check if you call strlen. 
If you find a smarter algorithm that passes the CI build, I can reconsider

Reply:
You only pass the max length of output string (`dest` in this case) but no need for input string. An example can be `llama_load_model_from_file(const char * path_model)`, there is no need for `path_mode_size`

If the compiler don't want `strlen`, let's go full cpp then

```cpp
int llama_split_prefix(char * dest, const char * split_path, int split_no, int split_count) {
    std::string str_split_path(split_path);
    char postfix[128];
    sprintf(postfix, "-%05d-of-%05d.gguf", split_no, split_count);
    std::string str_postfix(postfix);

    // check if dest ends with postfix
    auto size_prefix = str_split_path.size() - str_postfix.size();
    if (size_prefix > 0 && str_split_path.find(str_postfix, size_prefix) != std::string::npos) {
        strncpy(dest, split_path, size_prefix);
        return size_prefix;
    }

    return 0;
}
```

Reply:
Please read about `-Wstringop-overflow`, I always prefer to use strncpy. Show a CI run success with your changes, then I can consider to integrate it.

Reply:
I got CI build on my fork without problems: https://github.com/ngxson/llama.cpp/actions/runs/8382415859?pr=4

Here what you should add is `dest_len`, but not `split_path`, because that follows the same pattern with all other functions in `llama.h` (specify maxlen for output, no need to specify length for string input).

If this function is outside of `llama.h` then that won't be a problem, but now since it's part of the core library, I think it's important to have the same pattern across all functions.

Reply:
Good start, but you are not checking split_no and split_count

Reply:
I don’t understand what check you’re refering to. My implementation check if the path ends with an expected prefix or not - I don’t bother using sscanf for that reason.
Can you please give a case where my function fails?

Reply:
Sorry I read it too quickly. I will check tomorrow, thanks

Reply:
Yeah no worries. let me know if you need help. Bonne nuit !

Reply:
Thanks I applied your version + switching `split_path_len` to dest `maxlen`. Added you as co-author in the commit. Please confirm e474e456ebaa5a169d7ea6d12ddb9a7c4087d971

Reply:
It's LGTM, thanks!